### ISSUE TITLE
DeviceChild.Device members not disposed

### ISSUE THREAD

**Issue by OndrejPetrzilka** [sharpdx/SharpDX#718]
Example code:
```
using(Buffer buffer = new Buffer(...))
{
  var context = buffer.Device.ImmediateContext;
}
```
When `using `scope ends, buffer gets disposed, so far ok.
In buffer.Dispose(), DisposeDevice() is called.
```
private void DisposeDevice()
{
	if (this.Device__ != null)
	{
		((IUnknown)this.Device__).Release(); // Problem
		this.Device__ = null;
	}
}
```
Device gets released, but not any ComObjects it's holding, e.g. ImmediateContext.

The problem is, there's no way to use and deterministically release ImmediateContext obtained from DeviceChild. This may apply to other ComObjects as well.

So I end up not just passing buffer around, but also device.
Manually disposing buffer.Device is not a solution...when you consider that buffer is constructed in different places and buffer.Device is used multiple times.

Wouldn't it be OK to call `this.Device__.Dispose()` instead of `Release()`
...I'm not sure of consequences, that's why I'm asking

**Comment by Ziriax** [comment]
This seems related to #973 

Calling `this.Device__.Dispose()` would break existing code that references the `Device` wrapper instance after the `DeviceChild` wrapper is disposed, since the `NativePointer` is set to zero when disposed. One could argue that this is correct behavior, since one really disposes the managed wrapper, not the COM object itself, and the `DeviceChild` 'owns' its `Device` wrapper, so disposing the child must dispose the `Device` wrapper too... Only @xoofx can answer this.

With the current code another problem occurs when the `Device` wrapper that the `DeviceChild` created is eventually `Disposed` when `Configuration.EnableReleaseOnFinalizer` is enabled, then the device is released too much, resulting in a crash. So it feels to me that calling `Dispose` as you suggest is a valid solution, even when this would break existing code... Or some kind of 'soft-dispose' or `Detach` is needed, that calls `Release` once and only once, but keeps the `NativePointer`... This would keep existing code working.

The way I see this: since getting the `DeviceChild.Device.ImmediateContext` property always creates a new `Device` and `ImmediateContext` wrapper - bumping the reference of both the  `Device` and the `DeviceContext` - `DeviceChild.DisposeDevice` should not just call `Release` (or the propsed `Detach`) on the `Device`, but also on the `ImmediateContext`. When introducing `Device.Detach`, it could do just that, calling `ImmediateContext.Detach`.

The easiest solution would have been a convention in SharpDX: a managed wrapper represents a single reference count bump of a native object, and a managed wrapper is always owned by the class that creates it. This means that all explicit calls to `AddReference` and `Release` should be removed in the code (it only exists in a couple of places), and that wrapper instances should be created instead. This would also make the garbage collector work nicely (although since the GC doesn't know about video memory pressure, it is always a bad idea to use it IMHO, unless we patch the .NET GC to make it video memory aware 😉 )

Another solution would be to use a `GCHandle` to the managed `Device`, and set this as private data on the native device. When `DeviceChild.Device` is requested, then the existing managed `Device` wrapper is returned. The device is not touched when the `DeviceChild` is released. When the `Device` is disposed then the private data is cleared (maybe double checking the `GCHandle` refers to the `Device` being disposed). @xoofx must have had reasons not to do this beyond my knowledge of SharpDX internals.

PS:  Too bad `C#` doesn't support RAII with `struct` destructors, otherwise we could make real smart COM pointers I guess...

**Comment by xoofx** [comment]
>   var context = buffer.Device.ImmediateContext;

Yeah, this part in SharpDX is messy. When I was using SharpDX, the general rule I used was to never access the Device of the DeviceChild. Always use a singleton in your app. 😅 

`Configuration.EnableReleaseOnFinalizer` was to allow a smoother transition from SlimDX.

I don't remember exactly the details, but it might be ok to call dispose on the Device, as long as you don't keep a reference to the Device (which could happen, but it is very hopefully uncommon). Keeping reference here is indeed possible in C++, but I would not mind breaking it in C#. I would prefer this rather than introducing a third Attach/Detach semantic.

If I had to redesign SharpDX, I would use plain structs and explicit/manual dispose all other the places.

PR welcome (for `this.Device__.Dispose()` instead of `Release()`)... just double check as much as possible the usages, as I don't recall all the details.

**Comment by Ziriax** [comment]
@xoofx  Thanks for the quick feedback, you must be very busy, so I really appreciate the time you take for SharpDX

Maybe we could start by marking `DeviceChild.Device` `[Obsolete]`? 

I will make a PR where a new object is made for each wrapper, so that `AddReference` and `Release` are not called anymore, so that `Dispose` can always be called. This will break some existing code, but fixes this issue and #973, and makes the GC work correctly (the latter could be nice to have when writing quick and dirty prototyping code).

I would prefer to add some unit tests that first reproduces the issues...
